### ISSUE TITLE
Marginal effects equal to zero in mlogit()

### ISSUE THREAD

**Issue by MrJerryTAO** [vincentarelbundock/marginaleffects#719]
Hi @vincentarelbundock, I found the reason for and potential solution to your note regarding mlogit() models #159 
> Note that the slopes function will always return estimates of zero for regressors before the vertical bar in the formula. This is because the [predict()](https://rdrr.io/r/stats/predict.html) function supplied by the mlogit package does not produce different predictions for different values of these variables.

This is an undesired result, as changes in predictors should make a difference and have marginal effects, otherwise they will not be called _predictors_. Then, the important question is compared to what a change should be assessed. 

Using an example from 'Run examples' link in mlogit::mlogit help document, we see that such predictors are both individual- and alternative-specific. They are also referred to as choice-specific variables, as the same person can face different attribute values of different choices. 

```r
library(mlogit)
data("Fishing", package = "mlogit")
Fish <- dfidx(Fishing, varying = 2:9, shape = "wide", choice = "mode")
summary(Model <- mlogit( ## a pure "conditional" model
  mode ~ price + catch, data = Fish)) # beach as reference by default
"                      Estimate Std. Error  z-value  Pr(>|z|)    
(Intercept):boat     0.8713749  0.1140428   7.6408 2.154e-14 ***
(Intercept):charter  1.4988884  0.1329328  11.2755 < 2.2e-16 ***
(Intercept):pier     0.3070552  0.1145738   2.6800 0.0073627 ** 
price               -0.0247896  0.0017044 -14.5444 < 2.2e-16 ***
catch                0.3771689  0.1099707   3.4297 0.0006042 ***"
avg_slopes(Model) # all nearly 0
avg_comparisons(Model) # all nearly 0
Fish[1:8, ] # price and catch are mode-specific within first 2 individuals
"   mode   income   price  catch    idx
1 FALSE 7083.332 157.930 0.0678 1:each
2 FALSE 7083.332 157.930 0.2601 1:boat
3  TRUE 7083.332 182.930 0.5391 1:rter
4 FALSE 7083.332 157.930 0.0503 1:pier
5 FALSE 1250.000  15.114 0.1049 2:each
6 FALSE 1250.000  10.534 0.1574 2:boat
7  TRUE 1250.000  34.534 0.4671 2:rter
8 FALSE 1250.000  15.114 0.0451 2:pier"
predict(Model, newdata = Fish[1:8, ])
"      beach      boat   charter      pier
1 0.1282391 0.3295740 0.3690034 0.1731834
2 0.1214869 0.3317954 0.3852498 0.1614679"
predict(Model, newdata = Fish[1:8, -1]) # newdata must include y column? 
"Error in model.frame.default(terms(
  formula, lhs = lhs, rhs = rhs, data = data,  : 
  object is not a matrix"
predict(Model, newdata = as.data.frame(Fish[1:8, -1])) # worked
```
Changing these alternative-specific predictor at the individual level altogether, so that each alternative receives the same change within the same individual, will results in no change in prediction. This is because a baseline-category choice model is identified only from the relative utility perspective: How the utility of each alternative (represented by alt.subset = ... in mlogit()) differs from the reference choice (represented by reflevel = ... in mlogit()) within each individual. 
```r
predict( # if price at all places increase by the same amount, prob unchanged
  # diff in price cross different alternatives within individual unchanged
  # only the difference in predictors from the reference level matters
  Model, newdata = Fish[1:8, ] %>% mutate(price = price + 1))
"      beach      boat   charter      pier
1 0.1282391 0.3295740 0.3690034 0.1731834
2 0.1214869 0.3317954 0.3852498 0.1614679"
```
If we instead change these alternative-specific predictor at the alternative level, so that the price attribute corresponding to one alternative changes while those to other alternatives are held constant, adjusted predictions do differ between alternatives for each individual. This is what marginal effects should measure - marginal effects need to be alternative-specific for alternative-specific predictors. See the adjusted predictions from predict() below. 
```r
predict( # if fishing price at beach increase, others as before, prob changed
  Model, newdata = Fish[1:8, ] %>% mutate(
    price = ifelse(idx$id2 == "beach", price + 1, price)))
"      beach      boat   charter      pier
1 0.1254932 0.3306121 0.3701657 0.1737289
2 0.1188659 0.3327853 0.3863992 0.1619497"
predict( # if fishing price by boat increase, others as before, prob changed
  Model, newdata = Fish[1:8, ] %>% mutate(
    price = ifelse(idx$id2 == "boat", price + 1, price)))
"      beach      boat   charter      pier
1 0.1292824 0.3241200 0.3720053 0.1745923
2 0.1224819 0.3263225 0.3884052 0.1627904"
predict( # if fishing price by charter increase, others as before, prob changed
  Model, newdata = Fish[1:8, ] %>% mutate(
    price = ifelse(idx$id2 == "charter", price + 1, price)))
"      beach      boat   charter      pier
1 0.1294083 0.3325789 0.3632504 0.1747624
2 0.1226437 0.3349549 0.3793958 0.1630055"
predict( # if fishing price at pier increase, others as before, prob changed
  Model, newdata = Fish[1:8, ] %>% mutate(
    price = ifelse(idx$id2 == "pier", price + 1, price)))
"      beach      boat   charter      pier
1 0.1287852 0.3309775 0.3705748 0.1696625
2 0.1219691 0.3331124 0.3867789 0.1581396"
```
Compared to the no-change block of prediction results, the above examples show that predictions for all four alternatives, including the reference category, are different now. So, my suggestion is that data frame building in {marginaleffects} needs to adjust predictor values on both individual and alternative levels, whereas the current adjustment is only made on the individual level and ignores alternative-level adjustments. Perhaps the efficient coding has a form like this
```r
Fish %>% # a more efficient?? programming
  group_by(idx$id1) %>% # so the predictor adjustment is within each indv
  mutate(price = case_when( # dfidx objects are special data frames
    idx$id2 == "beach" ~ price + 1, # it has idx as a list of two columns
    TRUE ~ price # id1 is individual index, id2 is alternative index
  ))
```

**Comment by MrJerryTAO** [comment]
Adding to the examples above, mlogit() allows alternative-specific variables to have alternative-specific coefficients. I allowed the effects of price on utility to differ based on for which alternative the fee was assessed; this suggests that a $1 price increase deterred customers more if the tentative fishing location is a pier and less if the tentative fishing is with a charter group. Currently {marginaleffects} does not work well with dfidx data frames as they include a variable as a list of two lists of indices. However, transforming dfidx data frames to regular data frames will work for both mlogit() and {marginaleffects}. 

```r
summary(Model <- mlogit( # alternative-specific coefficient
  mode ~ price:idx$id2 + catch | 1, data = Fish)) # beach as reference
"                      Estimate Std. Error  z-value  Pr(>|z|)    
(Intercept):boat     0.2248294  0.1667979   1.3479 0.1776859    
(Intercept):charter  0.6051447  0.1877563   3.2230 0.0012684 ** 
(Intercept):pier     0.4391349  0.1563844   2.8080 0.0049843 ** 
catch                0.3813461  0.1103155   3.4569 0.0005465 ***
price:idx$beach     -0.0350644  0.0031909 -10.9889 < 2.2e-16 ***
price:idx$boat      -0.0221161  0.0019959 -11.0808 < 2.2e-16 ***
price:idx$charter   -0.0198261  0.0018636 -10.6387 < 2.2e-16 ***
price:idx$pier      -0.0390785  0.0032029 -12.2011 < 2.2e-16 ***"
avg_slopes(Model) # list extractor not allowed inside formula
"Error: Class of the `idx` variable is class is not supported.
In addition: Warning messages:
1: In is.recursive(.$object) && !is.primitive(.$object) && n > 0 :
  'length(x) = 2 > 1' in coercion to 'logical(1)'
2: Using `$` in model formulas can produce unexpected results.
  Specify your model using the `data` argument instead.
  Try: mode
  ~ price + catch:id2 | 1, data = idx 
3: These variables were not found: id2.  Try specifying the
  `newdata` argument explicitly and make sure the missing variable
  is included."
avg_comparisons(Model) # similar error
"Error: Class of the `idx` variable is class is not supported.
In addition: Warning messages:
1: Using `$` in model formulas can produce unexpected results.
  Specify your model using the `data` argument instead.
  Try: mode
  ~ price + catch:id2 | 1, data = idx 
2: These variables were not found: id2.  Try specifying the
  `newdata` argument explicitly and make sure the missing variable
  is included."
```
The above model is equivalent to the following one without using $ in formula. This avoid errors in avg_slopes() in avg_comparisons() but I do not think they give marginal effects for Term price across four alternatives in Group.
```r
summary(Model <- mlogit( # alternative-specific coefficient
  mode ~ catch | 1 | price, data = Fish))
avg_slopes(Model)
"   Group  Term  Estimate Std. Error         z Pr(>|z|)     2.5 %
 beach   catch  1.27e-15   8.02e-10  1.59e-06   1.0000 -1.57e-09
 boat    catch  1.58e-14   2.20e-09  7.20e-06   1.0000 -4.31e-09
 charter catch -3.06e-14   1.64e-09 -1.87e-05   1.0000 -3.21e-09
 pier    catch  7.32e-15   8.94e-10  8.19e-06   1.0000 -1.75e-09
 beach   price -5.16e-04   2.09e-04 -2.47e+00   0.0134 -9.25e-04
 boat    price  4.23e-04   2.94e-04  1.44e+00   0.1502 -1.53e-04
 charter price  1.34e-03   2.81e-04  4.78e+00   <0.001  7.92e-04
 pier    price -1.25e-03   2.47e-04 -5.06e+00   <0.001 -1.73e-03"
avg_comparisons(Model) # Slightly different from avg_slopes, but why? 
"   Group  Term Contrast  Estimate Std. Error         z Pr(>|z|)
 beach   catch       +1  2.54e-18   1.62e-13  1.56e-05   1.0000
 boat    catch       +1 -8.91e-18   1.03e-12 -8.67e-06   1.0000
 charter catch       +1  3.34e-18   7.86e-13  4.24e-06   1.0000
 pier    catch       +1  1.68e-19   4.01e-13  4.19e-07   1.0000
 beach   price       +1 -5.16e-04   2.09e-04 -2.47e+00   0.0135
 boat    price       +1  4.23e-04   2.94e-04  1.44e+00   0.1500
 charter price       +1  1.34e-03   2.81e-04  4.79e+00   <0.001
 pier    price       +1 -1.25e-03   2.47e-04 -5.06e+00   <0.001"
```
With alternative-specific coefficients (4 coefficients of price, including one for the reference category) of alternative-specific variables (price), BOTH different coefficient sizes (across 4 alternatives including the reference category) AND predictor value differences from the reference category (3 values of difference for each individual) matter when making predictions. Therefore, predictions will change if we add $1 to an original price both (1) for all alternatives altogether, which is solely due to variation in coefficients across alternatives, as changes in predictor values cancel when taking differences; and (2) for one of the alternatives holding price of other alternatives constant, which is due to one coefficient alone. I do not think what Approach (1) gives should be marginal effects, and I favor Approach (2) as marginal effects. See the adjusted predictions from both approaches according to predict() below. 
```r
predict(Model, newdata = Fish[1:8, ])
"       beach      boat   charter       pier
1 0.03697108 0.3849901 0.5478148 0.03022411
2 0.16813280 0.2890489 0.3028721 0.23994612"
predict( # if price at all places increase, prob changed
  Model, newdata = Fish[1:8, ] %>% mutate(price = price + 1))
"       beach      boat   charter       pier
1 0.03648553 0.3848854 0.5489214 0.02970768
2 0.16688842 0.2906488 0.3052467 0.23721610"
predict( # if fishing price at beach increase, others as before, prob changed
  Model, newdata = Fish[1:8, ] %>% mutate(
    price = ifelse(idx$id2 == "beach", price + 1, price)))
"       beach      boat   charter       pier
1 0.03574271 0.3854811 0.5485135 0.03026266
2 0.16328545 0.2907333 0.3046370 0.24134430"
predict( # if fishing price by boat increase, others as before, prob changed
  Model, newdata = Fish[1:8, ] %>% mutate(
    price = ifelse(idx$id2 == "boat", price + 1, price)))
"       beach      boat   charter       pier
1 0.03728505 0.3797671 0.5524671 0.03048079
2 0.16920257 0.2845254 0.3047992 0.24147282"
predict( # if fishing price by charter increase, others as before, prob changed
  Model, newdata = Fish[1:8, ] %>% mutate(
    price = ifelse(idx$id2 == "charter", price + 1, price)))
"       beach      boat   charter       pier
1 0.03737299 0.3891753 0.5428991 0.03055267
2 0.16913843 0.2907778 0.2987025 0.24138129"
predict( # if fishing price at pier increase, others as before, prob changed
  Model, newdata = Fish[1:8, ] %>% mutate(
    price = ifelse(idx$id2 == "pier", price + 1, price)))
"       beach      boat   charter       pier
1 0.03701395 0.3854365 0.5484501 0.02909948
2 0.16969328 0.2917317 0.3056832 0.23289187"
```
It appears predictions() in {marginaleffects} takes Approach (1) for any type of predictors. For individual-specific predictors like income, Approach (1) is okay and should be used, so that the income of one individual does not vary across different alternatives within the same individual even with marginal changes. However, for alternative-specific predictors like price, Approach (1) is wrong,  and Approach (2) must be used instead, so that the price of different alternatives gets changed separately within each individual. 

"Holding all else constant" should mean that when the price for alternative "beach" changes, prices for other alternatives like "boat" remain unchanged. Therefore, although this relaxed model specification appears different from the constrained case of one-coefficient-for-all-alternatives of an alternative-specific variable in my previous comment https://github.com/vincentarelbundock/marginaleffects/issues/159#issuecomment-1464965004, mlogit() with alternative-specific variables with alternative-specific coefficients may not need a special treatment in {marginaleffects} different from mlogit() models with alternative-specific variables with constant coefficients. For both situations, finding alternative-specific marginal effects is the key. The root is data frame building, which must adjust predictor values alternative by alternative within each individual.

**Comment by MrJerryTAO** [comment]
The output of effects() in {mlogit} that gives marginal effects of the covariates confirmed my thoughts that "marginal effects need to be alternative-specific for alternative-specific predictors." From effects.mlogit {mlogit} help document
> If the covariate is alternative specific, a J×J matrix is returned, J being the number of alternatives. Each line contains the marginal effects of the covariate of one alternative on the probability to choose any alternative. If the covariate is individual specific, a vector of length J is returned.
See the output of example scripts. 
```r
summary(Model <- mlogit( # alternative-specific coefficient
  mode ~ catch | income | price, data = Fish))
"                       Estimate  Std. Error  z-value  Pr(>|z|)    
(Intercept):boat     2.1208e-01  2.3183e-01   0.9148 0.3602839    
(Intercept):charter  1.1107e+00  2.3565e-01   4.7133 2.438e-06 ***
(Intercept):pier     7.8655e-01  2.2355e-01   3.5184 0.0004342 ***
catch                3.7243e-01  1.1022e-01   3.3791 0.0007273 ***
income:boat         -1.1363e-05  5.5927e-05  -0.2032 0.8389908    
income:charter      -1.7384e-04  5.5361e-05  -3.1402 0.0016886 ** 
income:pier         -1.1734e-04  5.3489e-05  -2.1937 0.0282546 *  
price:beach         -3.6787e-02  3.3379e-03 -11.0211 < 2.2e-16 ***
price:boat          -2.3699e-02  2.1554e-03 -10.9951 < 2.2e-16 ***
price:charter       -1.8377e-02  1.8943e-03  -9.7014 < 2.2e-16 ***
price:pier          -3.8193e-02  3.2480e-03 -11.7590 < 2.2e-16 ***"
avg_slopes(Model) # Incorrect?
"   Group   Term  Estimate Std. Error         z Pr(>|z|)     2.5 %
 beach   catch   2.19e-15   2.78e-07  7.86e-09  1.00000 -5.46e-07
 boat    catch   3.71e-14   4.74e-07  7.83e-08  1.00000 -9.30e-07
 charter catch  -8.76e-15   3.26e-07 -2.68e-08  1.00000 -6.40e-07
 pier    catch  -1.23e-14   3.52e-07 -3.48e-08  1.00000 -6.89e-07
 beach   income  9.05e-06   3.75e-06  2.41e+00  0.01588  1.69e-06
 boat    income  2.69e-05   5.62e-06  4.79e+00  < 0.001  1.59e-05
 charter income -3.07e-05   6.06e-06 -5.06e+00  < 0.001 -4.26e-05
 pier    income -5.25e-06   4.54e-06 -1.16e+00  0.24700 -1.41e-05
 beach   price  -6.80e-04   2.21e-04 -3.07e+00  0.00215 -1.11e-03
 boat    price  -1.44e-04   3.18e-04 -4.53e-01  0.65087 -7.67e-04
 charter price   1.96e-03   3.02e-04  6.49e+00  < 0.001  1.37e-03
 pier    price  -1.14e-03   2.62e-04 -4.33e+00  < 0.001 -1.65e-03"
NewData <- with(Fish, data.frame(
  price = tapply(price, idx(Model, 2), mean),
  catch = tapply(catch, idx(Model, 2), mean),
  income = mean(income)))
effects(Model, covariate = "income", data = NewData)
"        beach          boat       charter          pier 
 2.980132e-06  3.665677e-05 -3.891425e-05 -7.226567e-07 "
effects(Model, covariate = "catch", type = "aa", data = NewData)
"                beach         boat      charter          pier
beach    0.0111181369 -0.004924551 -0.005790897 -0.0004026897
boat    -0.0049245433  0.091246140 -0.080709207 -0.0056123896
charter -0.0057908864 -0.080709184  0.093099811 -0.0065997411
pier    -0.0004026897 -0.005612398 -0.006599753  0.0126148402"
effects(Model, covariate = "price", type = "aa", data = NewData)
"                beach          boat       charter          pier
beach   -1.098183e-03  0.0004864174  0.0005719898  3.977526e-05
boat     3.133574e-04 -0.0058061534  0.0051356697  3.571263e-04
charter  2.857415e-04  0.0039824581 -0.0045938526  3.256531e-04
pier     4.129581e-05  0.0005755511  0.0006768043 -1.293651e-03"
effects( # IIA property: other alternative gets proportionally changed
  Model, covariate = "catch", type = "rr", data = NewData)
"               beach         boat      charter         pier
beach    0.086996145 -0.002764755 -0.002764755 -0.002764755
boat    -0.027373943  0.036392173 -0.027373943 -0.027373943
charter -0.118325998 -0.118325998  0.116071825 -0.118325998
pier    -0.002120876 -0.002120876 -0.002120876  0.058296796"
effects( # IIA holds for prediction w/ alternative-specific coefficients
  Model, covariate = "price", type = "rr", data = NewData)
"             beach       boat    charter       pier
beach   -3.6873809  0.1171857  0.1171857  0.1171857
boat     0.5621528 -0.7473517  0.5621528  0.5621528
charter  0.7827795  0.7827795 -0.7678671  0.7827795
pier     0.1386594  0.1386594  0.1386594 -3.8113497"
```

**Comment by vincentarelbundock** [comment]
Thanks. I will change the vignette to explain that alternative-specific effects are not supported.

The whole `marginaleffects` infrastructure is designed to change all rows of a dataset in the same way. Supporting alternative-specific changes to specific rows would be *a lot* of work.

Unfortunately, I do not have the time for that right now.

**Comment by MrJerryTAO** [comment]
> The whole `marginaleffects` infrastructure is designed to change all rows of a dataset in the same way. Supporting alternative-specific changes to specific rows would be _a lot_ of work.

There might be a short cut: Consistently use the "wide" data format internally to generate counterfactual data when working with mlogit() and other discrete choice models with alternative-specific effects. In that way, alternative-specific predictors like price is stored in separate columns with different variable names (e.g., price.beach, ..., price.pier), and your existing frame of work may find it easy to change all rows in each column incrementally without altering other columns. 

mlogit() muse use the "long" data format for fitting; in the returned model object, object$model contains the data frame of class ‘dfidx_mlogit’, ‘dfidx’ and 'data.frame', which is also in the long format; predict.mlogit() also only recognize a long format data. However, mlogit.data() easily transforms a wide format data to long, producing a ‘dfidx’ data frame. So the solution might be (1) extract the long-formatted data from object$model, transform it to the wide format (using pivot_wider {tidyr} given the model formula I guess?); (2) make datagrid() adjustements to the wide data; (3) transform the data to long using mlogit.data(); (4) make predictions using predict.mlogit().

**Comment by vincentarelbundock** [comment]
This is a neat idea, but I think it understates how difficult this would be. The wide format is weird and would require special processing. Reshaping is possible but tricky to support all user input (and we don't use `tidyr` in this package).

It would take me many many hours of work to implement, so I am unlikely to work myself on this in the short or medium run.

Perhaps a contributor will submit code in a Pull Request eventually. I would potentially be open to merging such code, but I strongly suggest to potential contributors that they first open an Issue to describe exactly what the implementation plan is before writing code (e.g., which specific files and functions to modify, what the logic of the implementation is, etc.). I wouldn't want someone to put in a lot of efforts and then be unable to merge.

**Comment by MrJerryTAO** [comment]
> Important: The newdata argument for mlogit models must be a “balanced” data frame, that is, it must have a number of rows that is a multiple of the number of choices.

On another note, both mlogit() and predict.mlogit() are okay to take unbalanced data: If one row is missing, it means one of the alternatives (i.e. choices) is not applicable, impossible, or out of consideration for an individual, so the fitted or predicted probability is exactly 0. The following examples confirmed this, and found that fitting and predicting from unbalanced data are equivalent to those from balanced data with NA on the not-applicable choices. 

Balanced data for fitting mlogit() but imbalanced in predict()
```r
library(mlogit)
data("Fishing", package = "mlogit")
Fish <- dfidx(Fishing, varying = 2:9, shape = "wide", choice = "mode")
Fish <- mlogit.data( # same as above
  Fishing, varying = 2:9, shape = "wide", choice = "mode")
summary(Model <- mlogit( #
  mode ~ price + catch, data = Fish)) # beach as reference
"                      Estimate Std. Error  z-value  Pr(>|z|)    
(Intercept):boat     0.8713749  0.1140428   7.6408 2.154e-14 ***
(Intercept):charter  1.4988884  0.1329328  11.2755 < 2.2e-16 ***
(Intercept):pier     0.3070552  0.1145738   2.6800 0.0073627 ** 
price               -0.0247896  0.0017044 -14.5444 < 2.2e-16 ***
catch                0.3771689  0.1099707   3.4297 0.0006042 ***"
Fishing[1:2, ] # wide: first 2 persons
"     mode price.beach price.pier price.boat price.charter catch.beach
1 charter     157.930    157.930    157.930       182.930      0.0678
2 charter      15.114     15.114     10.534        34.534      0.1049
  catch.pier catch.boat catch.charter   income
1     0.0503     0.2601        0.5391 7083.332
2     0.0451     0.1574        0.4671 1250.000"
Fish[1:8, ] # price and catch are alt-specific within each individual
"   mode   income   price  catch    idx
1 FALSE 7083.332 157.930 0.0678 1:each
2 FALSE 7083.332 157.930 0.2601 1:boat
3  TRUE 7083.332 182.930 0.5391 1:rter
4 FALSE 7083.332 157.930 0.0503 1:pier
5 FALSE 1250.000  15.114 0.1049 2:each
6 FALSE 1250.000  10.534 0.1574 2:boat
7  TRUE 1250.000  34.534 0.4671 2:rter
8 FALSE 1250.000  15.114 0.0451 2:pier"
predict(Model, newdata = Fish[1:8, ]) 
"      beach      boat   charter      pier
1 0.1282391 0.3295740 0.3690034 0.1731834
2 0.1214869 0.3317954 0.3852498 0.1614679"
predict(Model, newdata = Fish[1:7, ]) # pier not considered for person 2
"      beach     boat   charter      pier
1 0.1282391 0.329574 0.3690034 0.1731834
2 0.1448804 0.395686 0.4594336 0.0000000"
predict(Model, newdata = Fish[1:5, ]) # only beach possible for person 2
"      beach     boat   charter      pier
1 0.1282391 0.329574 0.3690034 0.1731834
2 1.0000000 0.000000 0.0000000 0.0000000"
```

Imbalanced data for fitting mlogit()
```
summary(Model <- mlogit( # remove pier for person 1 & beach for person 2
  mode ~ price + catch, data = Fish[-c(4, 5), ])) 
Model$model[1:8, ] # missing rows recreated with NA
"   mode   price  catch    idx probabilities     linpred
1 FALSE 157.930 0.0678 1:each     0.1553338 -3.88889097
2 FALSE 157.930 0.2601 1:boat     0.3984929 -2.94677751
3  TRUE 182.930 0.5391 1:rter     0.4461733 -2.83375996
4    NA      NA     NA 1:pier     0.0000000          NA
5    NA      NA     NA 2:each     0.0000000          NA
6 FALSE  10.534 0.1574 2:boat     0.3775446  0.66786191
7  TRUE  34.534 0.4671 2:rter     0.4383708  0.81723826
8 FALSE  15.114 0.0451 2:pier     0.1840846 -0.05043168"
summary(Model <- mlogit( # identical result as above
  mode ~ price + catch, data = Fish %>% 
    mutate(price = case_when(
      idx$id1 == 1 & idx$id2 == "pier" ~ NA,
      idx$id1 == 2 & idx$id2 == "beach" ~ NA,
      TRUE ~ price)) )) 
Model$model[1:8, ] # missing rows also recreated with NA
"   mode   price  catch    idx probabilities     linpred
1 FALSE 157.930 0.0678 1:each     0.1553338 -3.88889097
2 FALSE 157.930 0.2601 1:boat     0.3984929 -2.94677751
3  TRUE 182.930 0.5391 1:rter     0.4461733 -2.83375996
4    NA      NA     NA 1:pier     0.0000000          NA
5    NA      NA     NA 2:each     0.0000000          NA
6 FALSE  10.534 0.1574 2:boat     0.3775446  0.66786191
7  TRUE  34.534 0.4671 2:rter     0.4383708  0.81723826
8 FALSE  15.114 0.0451 2:pier     0.1840846 -0.05043168"
predict(Model, newdata = Fish[1:8, ]) # balanced prediction data
"      beach      boat   charter      pier
1 0.1283961 0.3293870 0.3687987 0.1734181
2 0.1216390 0.3316205 0.3850478 0.1616927"
predict(Model, newdata = Fish[1:7, ]) # pier not considered for person 2
"      beach      boat   charter      pier
1 0.1283961 0.3293870 0.3687987 0.1734181
2 0.1451007 0.3955834 0.4593158 0.0000000"
predict(Model, newdata = Fish[1:8, ] %>% mutate(price = ifelse(
  idx$id1 == 2 & idx$id2 == "pier", NA, price))) # same result
```
So, it might be reasonable to allow predictions() and other marginaleffects functions to take imbalanced data, by filling the missing rows with NA in predictors if necessary. expand() in tidyr might be relevant https://tidyr.tidyverse.org/reference/expand.html. 
```r
predictions(Model)[1:8, ] # same as fitted()
"   Group Estimate Std. Error    z Pr(>|z|) 2.5 % 97.5 %
 beach      0.155     0.0144 10.8   <0.001 0.127  0.183
 boat       0.398     0.0168 23.7   <0.001 0.365  0.432
 charter    0.446     0.0167 26.7   <0.001 0.413  0.479
 pier       0.000         NA   NA       NA    NA     NA
 beach      0.000         NA   NA       NA    NA     NA
 boat       0.378     0.0159 23.7   <0.001 0.346  0.409
 charter    0.438     0.0167 26.2   <0.001 0.406  0.471
 pier       0.184     0.0150 12.2   <0.001 0.155  0.214"
predictions(Model, newdata = Fish[1:8, ]) # does not work well with dfidx
"Error: Unable to compute predicted values with this model. You can try
  to supply a different dataset to the `newdata` argument. This
  error was also raised:
  attempt to select less than one element in get1index"
predictions(Model, newdata = Fish)[1:8, ] # same error
predictions(Model, newdata = Fish[1:8, ] %>% as.data.frame()) # worked
"  Group Estimate Std. Error    z Pr(>|z|) 2.5 % 97.5 % income price
 beach      0.128     0.0115 11.1   <0.001 0.106  0.151   7083 157.9
 boat       0.329     0.0155 21.2   <0.001 0.299  0.360   7083 157.9
 charter    0.369     0.0156 23.7   <0.001 0.338  0.399   7083 182.9
 pier       0.173     0.0135 12.8   <0.001 0.147  0.200   7083 157.9
 beach      0.122     0.0110 11.1   <0.001 0.100  0.143   1250  15.1
 boat       0.332     0.0150 22.2   <0.001 0.302  0.361   1250  10.5
 charter    0.385     0.0161 24.0   <0.001 0.354  0.417   1250  34.5
 pier       0.162     0.0129 12.6   <0.001 0.136  0.187   1250  15.1"
predictions(Model, newdata = Fish[ # 
  1:8, ] %>% as.data.frame() %>% mutate(price = ifelse(
    idx$id1 == 2 & idx$id2 == "pier", NA, price))) 
"   Group Estimate Std. Error    z Pr(>|z|) 2.5 % 97.5 % income price
 beach      0.128     0.0115 11.1   <0.001 0.106  0.151   7083 157.9
 boat       0.329     0.0155 21.2   <0.001 0.299  0.360   7083 157.9
 charter    0.369     0.0156 23.7   <0.001 0.338  0.399   7083 182.9
 pier       0.173     0.0135 12.8   <0.001 0.147  0.200   7083 157.9
 beach      0.145     0.0135 10.7   <0.001 0.119  0.172   1250  15.1
 boat       0.396     0.0163 24.3   <0.001 0.364  0.428   1250  10.5
 charter    0.459     0.0169 27.1   <0.001 0.426  0.493   1250  34.5
 pier       0.000         NA   NA       NA    NA     NA   1250    NA"
predictions(Model, newdata = Fish[1:5, ])
"Error: The `newdata` argument for `mlogit` models must be a data frame 
with a number of rows equal to a multiple of the number of choices: 4."
```

**Comment by vincentarelbundock** [comment]
I don't use the `tidyverse` in this package.

But if I understand correctly, all the user has to do is "complete" the dataset before feeding it to `newdata`. That seems pretty trivial, no?

**Comment by MrJerryTAO** [comment]
> That seems pretty trivial, no?
I think so.

**Comment by vincentarelbundock** [comment]
OK, then users can do that.

**Comment by vincentarelbundock** [comment]
@MrJerryTAO I think I found a nice solution for alternative-specific changes in numeric predictors. The development version of the website has an example where I increase the wait time only for air travel. See here:

https://vincentarelbundock.github.io/marginaleffects/dev/articles/categorical.html#mlogit-package

This used the dev version of the package from GitHub.

I know this is not ideal, but maybe it's not so bad either.

**Comment by MrJerryTAO** [comment]
@vincentarelbundock, excellent! I did not know that comparisons(variables = ) can be `Function which accepts a numeric vector and returns a data frame with two columns of "low" and "high" values to compare. See examples below.` I did not see examples in the help document. Then one would just need to write a for loop for all alternatives by each alternative-specific variable, with some meaningful small increment, and specify low = x - increment/2, high = x + increment/2, and eps=increment correctly, to retrieve marginal effect as slopes. 

It doesn't seem that slopes() can work in a similar way. But great finding and simple solution! Does comparisons() and slopes() otherwise calculate increments in the same way? I found in help that for eps argument, derivatives are `(f(x+eps)-f(x))/eps` for both slopes() and comparisons(), but slopes() details says (f(X+ε/2)−f(X+ε/2))/ε despite the incorrect + sign. 

I was thinking to instead specify something like `comparison = function(hi, lo) ifelse(mode == "air", 0, hi - lo)` perhaps by allowing such vectors to have row names involving alternative names. But it may not work, as I assume that by this stage, hi and lo are already predicted y probabilities (instead of adjusted x values) that would equal on each row of an alternative-specific variable. 

> Note that the slopes function will always return estimates of zero for regressors before the vertical bar in the formula.

Formula for mlogit could actually use two vertical bars. In `y ~ x | z | w`, x is an alternative-specific variable with one constant coefficient, z is an individual-specific variable with J - 1 alternative-specific coefficients, and w is an alternative-specific variable with J alternative-specific coefficients. This is equivalent to `y ~ x + w:alt | z` as used in earlier versions of mlogit https://www.rdocumentation.org/packages/mlogit/versions/0.1-0/topics/mlogit. 

Your proposed approach should get marginal effects in all cases. However, without your new approach, comparisons() and slopes() can still get non-zero estimates for w. but they are not marginal effects
